src/operator/tensor/matrix_op-inl.h

```diff
@@ -620,6 +626,227 @@ void SliceEx(const nnvm::NodeAttrs& attrs,
   }
 }
 
+template<int ndim>
+inline void GetIndexRange(const SliceParam& param,
+                          const TShape& dshape,
+                          common::StaticArray<int, ndim>* begin,
+                          common::StaticArray<int, ndim>* end,
+                          common::StaticArray<int, ndim>* step) {
+  CHECK_NE(dshape.ndim(), 0U);
+  CHECK_NE(dshape.Size(), 0U);
+  CHECK_LE(param.begin.ndim(), dshape.ndim())
+    << "Slicing axis exceeds data dimensions";
+  CHECK_LE(param.end.ndim(), dshape.ndim())
+    << "Slicing axis exceeds data dimensions";
+  CHECK_EQ(param.begin.ndim(), param.end.ndim())
+    << "begin and end must have the same length";
+  CHECK_EQ(ndim, dshape.ndim())
+    << "Static array size=" << ndim
+    << " is not equal to data shape ndim=" << dshape.ndim();
+
+  if (param.step.has_value()) {
+    CHECK_EQ(param.step.value().ndim(), param.begin.ndim())
+      << "step and begin must have the same length";
+  }
+
+  for (index_t i = 0; i < param.begin.ndim(); ++i) {
+    int b = 0, e = dshape[i], s = 1;
+    const int len = dshape[i];
+    if (param.step.has_value()) {
+      const auto& opt_step_val = param.step.value()[i];
+      if (opt_step_val.has_value()) {
+        s = opt_step_val.value();
+        CHECK_NE(s, 0) << "slice op step[" << i << "] cannot be 0";
+      }
+    }
+
+    if (param.begin[i].has_value()) {
+      b = param.begin[i].value();
+      if (b < 0) {
+        b += len;
+        CHECK_GE(b, 0) << "slicing with begin[" << i << "]="
+                       << b - len << " exceeds limit of " << len;
+      }
+    } else if (s < 0) {
+      b = len - 1;
+    }
+    CHECK_LT(b, len) << "slicing with begin[" << i << "]="
+                     << b << " exceends limit of " << len;
+
+    if (param.end[i].has_value()) {
+      e = param.end[i].value();
+      if (e < 0) {
+        e += len;
+        CHECK_GE(e, 0) << "slicing with end[" << i << "]="
+                       << e - len << " exceeds limit of " << len;
+      }
+    } else if (s < 0) {
+      e = -1;
+    }
+    CHECK_LE(e, len) << "slicing with end[" << i << "]="
```

Comment:
In thsse CHECK_XXX macros, is function call made for every <<, even if the check succeeds?  If so, does this affect performance?

Reply:
It's not called if check is successful. See the code:
https://github.com/dmlc/dmlc-core/blob/595d02c0e87be8a0846700462b6f45f1b1031e39/include/dmlc/logging.h#L89
All the << calls happen only when `_check_err` has a non-empty string stored.